archipelago/pipeline.py

```diff
@@ -0,0 +1,1028 @@
+                continue
+            path = []
+            curr_node = io_node
+            idx = 0
+            while curr_node not in graph.get_mems() and len(graph.sinks[curr_node]) < 2:
+                assert len(graph.sinks[curr_node]) == 1, len(graph.sinks[curr_node])
+                path.append((curr_node, idx))
+                curr_node = graph.sinks[curr_node][0]
+                idx += 1
+            break_crit_path(graph, id_to_name, path, placement, routing)
+
+def break_at(graph, node1, id_to_name, placement, routing):    
+    path = []
+    curr_node = node1
+    kernel = graph.get_node(curr_node).kernel
+
+    while len(graph.sinks[curr_node]) == 1:
+        if len(graph.sources[graph.sinks[curr_node][0]]) > 1 or graph.get_node(graph.sinks[curr_node][0]).kernel != kernel:
+            break
+        curr_node = graph.sinks[curr_node][0]
+
+    idx = 0
+    while len(graph.sources[curr_node]) == 1:
+        if len(graph.sinks[curr_node]) > 1 or graph.get_node(graph.sources[curr_node][0]).kernel != kernel:
+            break
+        path.append((curr_node, idx))
+        curr_node = graph.sources[curr_node][0]
+        if curr_node in graph.get_ponds():
+            break
+    
+    if curr_node in graph.get_ponds():        
+        print("\t\tFound pond for branch delay matching", curr_node)
+        for source in graph.sources[curr_node]:
+            if graph.get_node(source).port == "flush":
+                continue
+            graph.node_latencies[source] += 1
+        return
+
+    if len(path) == 0:
+        raise ValueError(f"Cant break at node: {node1}")
+    path.reverse()
+
+    ret = []
+
+    for p in path:
+        ret.append((p[0], idx))
+        idx += 1
+    break_crit_path(graph, id_to_name, ret, placement, routing)
+
+def add_delay_to_kernel(graph, kernel, added_delay, id_to_name, placement, routing):
+    kernel_output_nodes = graph.get_output_tiles_of_kernel(kernel)
+    for node in kernel_output_nodes:
+        for _ in range(added_delay):
+            break_at(graph, node, id_to_name, placement, routing)
+
+def branch_delay_match_all_nodes(graph, id_to_name, placement, routing):
+    nodes = graph.topological_sort()
+    node_cycles = {}
+
+    for node in nodes:
+        cycles = set()
+        g_node = graph.get_node(node)
+
+
+        if len(graph.sources[node]) == 0:
+            cycles = {0}
+
+        for parent in graph.sources[node]:
+            g_parent = graph.get_node(parent)
+            if parent not in node_cycles:
+                c = 0
+            else:
+                c = node_cycles[parent]
+            
+            if g_node.type_ == "route":
+                if graph.node_latencies[node] != 0:
+                    if len(graph.sinks[node]) > 0:
+                        child = graph.sinks[node][0]
+                        if (child not in id_to_name or "d_reg_" not in id_to_name[child]) and (parent not in graph.get_mems()):

```

Comment:
Not sure if it's a good way to test out names. The only thing you have control over is the ID name, which is maintained internally and not visible to normal users. That's why IDs are assigned based on their types. What if there is a name actually called `d_reg_`. Don't have to fix it this time but might need it fixed in the future.